cartographer_ros/cartographer_ros/node.cc

```diff
@@ -550,10 +563,36 @@ int Node::AddOfflineTrajectory(
       map_builder_bridge_.AddTrajectory(expected_sensor_ids, options);
   AddExtrapolator(trajectory_id, options);
   AddSensorSamplers(trajectory_id, options);
-  is_active_trajectory_[trajectory_id] = true;
   return trajectory_id;
 }
 
+bool Node::HandleGetTrajectoryStates(
+    ::cartographer_ros_msgs::GetTrajectoryStates::Request& request,
+    ::cartographer_ros_msgs::GetTrajectoryStates::Response& response) {
+  using TrajectoryState =
+      ::cartographer::mapping::PoseGraphInterface::TrajectoryState;
+  carto::common::MutexLocker lock(&mutex_);
+  response.status.code = ::cartographer_ros_msgs::StatusCode::OK;
+  response.trajectory_states.header.stamp = ros::Time::now();
+  for (const auto& entry : map_builder_bridge_.GetTrajectoryStates()) {
+    response.trajectory_states.trajectory_id.push_back(entry.first);
+    if (entry.second == TrajectoryState::ACTIVE) {
+      response.trajectory_states.trajectory_state.push_back(
+          ::cartographer_ros_msgs::TrajectoryStates::ACTIVE);
+    } else if (entry.second == TrajectoryState::FINISHED) {
+      response.trajectory_states.trajectory_state.push_back(
+          ::cartographer_ros_msgs::TrajectoryStates::FINISHED);
+    } else if (entry.second == TrajectoryState::FROZEN) {
+      response.trajectory_states.trajectory_state.push_back(
+          ::cartographer_ros_msgs::TrajectoryStates::FROZEN);
+    } else {
+      response.trajectory_states.trajectory_state.push_back(
+          ::cartographer_ros_msgs::TrajectoryStates::DELETED);
```

Comment:
You could warn if there is an unknown state rather than returning DELETED.
Or use switch, which will fail to compile if not all states are covered.